webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py

```diff
@@ -0,0 +1,224 @@
+
+    assert len(contexts) == 1
+    root_info = contexts[0]
+    assert len(root_info["children"]) == 1
+    child_info = root_info["children"][0]
+
+    # The ordering of fragmentNavigated events is not guaranteed between the
+    # root page and the iframe, find the appropriate events in the current list.
+    first_is_root = events[0]["context"] == root_info["context"]
+    root_event = events[0] if first_is_root else events[1]
+    child_event = events[1] if first_is_root else events[0]
+
+    assert_navigation_info(
+        root_event,
+        {"context": root_info["context"], "url": test_page_same_origin_frame}
+    )
+    assert_navigation_info(child_event, {"context": child_info["context"], "url": test_page})
+
+    remove_listener()
+
+
+async def test_cross_document(
+    bidi_session, new_tab, url, subscribe_events, wait_for_event
+):
+    await subscribe_events([FRAGMENT_NAVIGATED_EVENT])
+
+    on_frame_navigated = wait_for_event(FRAGMENT_NAVIGATED_EVENT)
+
+    result = await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=url(EMPTY_PAGE))
+
+    recursive_compare(
+        {
+            'context': new_tab["context"],
+            'navigation': result["navigation"],
+            'timestamp': any_int,
+            'url': url(EMPTY_PAGE)
+        },
+        await on_frame_navigated,
+    )
+
+
+@pytest.mark.parametrize(
+    "navigation_kind",
+    ["browsing_context.navigate", "script.call_function"],
+)
+@pytest.mark.parametrize(
+    "hash_before, hash_after",
+    [
+        ("", "#foo"),
+        ("#foo", "#bar"),
+        ("#foo", "#foo"),
+        ("#bar", ""),
+    ],
+    ids=[
+        "without hash to with hash",
+        "with different hashes",
+        "with identical hashes",
+        "with hash to without hash",
+    ],
+)
+async def test_history_api(
+    bidi_session, new_tab, url, subscribe_events, wait_for_event, hash_before, hash_after, navigation_kind
+):
+    await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=url(EMPTY_PAGE + hash_before), wait="complete"
+    )
+
+    await subscribe_events([FRAGMENT_NAVIGATED_EVENT])
+
+    on_frame_navigated = wait_for_event(FRAGMENT_NAVIGATED_EVENT)
+
+    if navigation_kind == "browsing_context.navigate":
+      result = await bidi_session.browsing_context.navigate(
+          context=new_tab["context"], url=url(EMPTY_PAGE + hash_after))
+
+      recursive_compare(
+          {
+              'context': new_tab["context"],
+              'navigation': result["navigation"],
+              'timestamp': any_int,
+              'url': url(EMPTY_PAGE + hash_after)
+          },
+          await on_frame_navigated,
+      )
+    elif navigation_kind == "script.call_function":
+      await bidi_session.script.call_function(
+          raw_result=True,
+          function_declaration="""(url) => {
+              history.pushState(null, null, url);
```

Comment:
Shouldn't we just use `document.location = "<new_url>"` to trigger a navigation for the evaluation tests? The History API specific tests I would propose to move to its own file given that it might become larger over time. Right now we only have a basic test.

Reply:
yeah, we could use document.location I think.